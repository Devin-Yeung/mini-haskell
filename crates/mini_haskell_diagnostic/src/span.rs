use ariadne::Label;

#[derive(Debug, PartialEq, Clone)]
pub struct Span {
    pub start: usize,
    pub end: usize,
}

impl From<logos::Span> for Span {
    fn from(value: logos::Span) -> Self {
        Span {
            start: value.start,
            end: value.end,
        }
    }
}

impl From<Span> for ariadne::Label {
    fn from(val: Span) -> Self {
        Label::new(val.start..val.end)
    }
}
