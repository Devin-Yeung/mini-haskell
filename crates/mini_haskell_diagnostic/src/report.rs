use crate::span::Span;

pub struct Label {
    pub span: Span,
    pub hint: String,
}

pub struct Report {
    pub message: String,
    pub offset: usize,
    pub labels: Vec<Label>,
}

pub struct ReportBuilder {
    pub message: Option<String>,
    pub offset: Option<usize>,
    pub labels: Vec<Label>,
}

impl ReportBuilder {
    pub fn new() -> Self {
        Self {
            message: None,
            offset: None,
            labels: Vec::new(),
        }
    }

    pub fn message(mut self, message: impl Into<String>) -> Self {
        self.message = Some(message.into());
        self
    }

    pub fn offset(mut self, offset: usize) -> Self {
        self.offset = Some(offset);
        self
    }

    pub fn label(mut self, span: Span, hint: impl Into<String>) -> Self {
        self.labels.push(Label {
            span,
            hint: hint.into(),
        });
        self
    }

    pub fn finish(self) -> Report {
        Report {
            message: self.message.unwrap(),
            offset: self.offset.unwrap(),
            labels: self.labels,
        }
    }
}
