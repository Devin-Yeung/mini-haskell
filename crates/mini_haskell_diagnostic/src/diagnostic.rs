use crate::report::Report;

pub trait Diagnostic {
    fn diagnose(&mut self, ctx: &mut DiagnosticContext);
}

pub struct DiagnosticContext {
    pub reports: Vec<Report>,
}

impl DiagnosticContext {
    pub fn new() -> Self {
        Self {
            reports: Vec::new(),
        }
    }

    pub fn report(&mut self, report: impl Into<Report>) {
        self.reports.push(report.into());
    }
}
