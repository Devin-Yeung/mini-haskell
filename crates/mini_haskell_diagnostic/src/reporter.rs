use crate::diagnostic::DiagnosticContext;
use ariadne::{ReportBuilder, Source};
use std::io::Cursor;
use std::ops::Range;

pub struct Reporter<'a> {
    // name of the source file
    // name: &'a str,
    // source code
    source: &'a str,
    // reporter config
    config: ReporterConfig,
    // diagnostic context
    context: DiagnosticContext,
}

impl<'a> Reporter<'a> {
    pub fn stringify(self) -> Vec<String> {
        self.context
            .reports
            .into_iter()
            .map(|report| {
                let mut builder: ReportBuilder<Range<usize>> =
                    ariadne::Report::build(ariadne::ReportKind::Error, (), report.offset)
                        .with_message(report.message)
                        .with_config(ariadne::Config::default().with_color(self.config.color));

                for label in report.labels {
                    builder = builder
                        .with_label(label.span.into())
                        .with_message(label.hint);
                }
                let report = builder.finish();

                let mut cursor = Cursor::new(Vec::<u8>::new());
                report
                    .write(ariadne::Source::from(self.source), &mut cursor)
                    .unwrap();
                String::from_utf8(cursor.into_inner()).unwrap()
            })
            .collect::<Vec<_>>()
    }
}

pub struct ReporterBuilder<'a> {
    source: Option<&'a str>,
    config: Option<ReporterConfig>,
    context: Option<DiagnosticContext>,
}

impl<'a> ReporterBuilder<'a> {
    pub fn new() -> Self {
        Self {
            source: None,
            config: None,
            context: None,
        }
    }

    pub fn source(mut self, source: &'a str) -> Self {
        self.source = Some(source);
        self
    }

    pub fn context(mut self, context: DiagnosticContext) -> Self {
        self.context = Some(context);
        self
    }

    pub fn config(mut self, config: ReporterConfig) -> Self {
        self.config = Some(config);
        self
    }
}

pub struct ReporterConfig {
    color: bool,
}
