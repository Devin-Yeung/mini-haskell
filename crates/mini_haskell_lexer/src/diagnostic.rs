#[cfg(test)]
mod test {
    use crate::error::LexingError;
    use crate::lexer::Token;
    use mini_haskell_diagnostic::diagnostic::diagnostic;
    use testsuite::unittest;

    unittest!(invalid_token, |path, _| {
        let result = diagnostic(path, |src| {
            Token::tokens(src)
                .into_iter()
                .filter(|tok| tok.is_err())
                .map(|tok| tok.err().unwrap())
                .collect::<Vec<LexingError>>()
        });

        insta::assert_snapshot!(result);
    });
}
