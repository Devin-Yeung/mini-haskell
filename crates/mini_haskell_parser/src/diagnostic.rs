#[cfg(test)]
mod test {
    use crate::parser::Parser;
    use mini_haskell_diagnostic::diagnostic::diagnostic;
    use testsuite::unittest;

    unittest!(unexpected_token, |path, _| {
        let result = diagnostic(path, |src| {
            let (_, errors) = Parser::parse(src);
            errors
        });

        insta::assert_snapshot!(result);
    });
}
