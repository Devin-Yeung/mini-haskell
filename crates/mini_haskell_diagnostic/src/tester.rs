use crate::diagnostic::DiagnosticContext;
use crate::report::Report;
use crate::reporter::{ReporterBuilder, ReporterConfig};

pub struct Tester<'a> {
    builder: ReporterBuilder<'a>,
}

impl<'a> Tester<'a> {
    pub fn new() -> Tester<'a> {
        let builder = ReporterBuilder::new().config(ReporterConfig { color: false });

        Tester { builder }
    }

    pub fn diagnose(self, source: &'a str, report: Vec<impl Into<Report>>) -> Vec<String> {
        let reporter = self
            .builder
            .source(source)
            .context(DiagnosticContext {
                reports: report.into_iter().map(|report| report.into()).collect(),
            })
            .build();
        reporter.stringify()
    }
}

#[cfg(test)]
mod tests {
    use crate::report::{Label, Report};
    use crate::span::Span;
    use crate::tester::Tester;
    use testsuite::unittest;

    unittest!(simple_diagnose, |src| {
        let reports = vec![Report {
            message: String::from("This is an example error"),
            offset: 0,
            labels: vec![Label {
                span: Span { start: 5, end: 6 },
                hint: "This is a hint".to_string(),
            }],
        }];
        let tester = Tester::new();
        let diagnostics = tester.diagnose(src, reports).join("\n");
        insta::assert_snapshot!(diagnostics);
    });
}
