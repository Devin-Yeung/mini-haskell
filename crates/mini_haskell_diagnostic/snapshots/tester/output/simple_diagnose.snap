---
source: crates/mini_haskell_diagnostic/src/tester.rs
expression: diagnostics
---
Error: This is an example error
   ╭─[<unknown>:1:1]
   │
 1 │ Hello World!
   │      ┬  
   │      ╰── This is a hint
───╯

